### PR TITLE
Toml universal implementor

### PR DESCRIPTION
**Pull request by kampffrosch94**
Adds toml universal implementor as a feature and an example.
I had to name the feature "toml_impl" because cargo doesn't allow a dependency and a feature to share the same name.

### CODE REVIEW

**Comment by phaazon** [comment]
> I had to name the feature "toml_impl" because cargo doesn't allow a dependency and a feature to share the same name.

Yeah, it was the same thing for `serde` when I introduced the `"json"` feature gate.

Thank you for the contribution! Let’s move that along and merge it whenever possible, and CI agrees! :)

**Comment by kampffrosch94** [comment]
CI fails because I made the toml feature optional.
You could let travis run cargo test --verbose --all-features
Or I could activate toml by default.
Which do you prefer?

**Comment by phaazon** [comment]
Hm, leave it this way, I’ll push some commits to fix the travis situation. Thanks!

File: Cargo.toml [hunk 1]
```diff
@@ -22,13 +22,15 @@ maintenance = { status = "actively-developed" }
 [features]
 default = ["json"]
 json = ["serde", "serde_json"]
+toml_impl = ["serde", "toml"]
```
Comment (phaazon) [inline Cargo.toml]: Just note for later: use that change to resort by name the features.

File: Cargo.toml [hunk 2]
```diff
@@ -22,13 +22,15 @@ maintenance = { status = "actively-developed" }
...
 arc = []
 
 [dependencies]
 any-cache = "0.2"
 notify = "4.0.3"
 serde = { version = "1", features = ["derive"], optional = true }
 serde_json = { version = "1", optional = true }
+toml = { version = "0.5.1", optional = true }
```
Comment (phaazon) [inline Cargo.toml]: @eijebong will be happy with this :D

File: examples/toml/main.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+struct Config {
+  msg: String,
+}
+
+fn main() {
+  let mut store: Store<(), SimpleKey> = Store::new(StoreOpt::default()).unwrap();
+  let ctx = &mut ();
+  // using cargo manifest dir to build our path lets us find our resource
+  // even if the cwd is not in the project root
+  let resource_path = if let Ok(manifest_dir) = env::var("CARGO_MANIFEST_DIR") {
```
Comment (phaazon) [inline examples/toml/main.rs]: Not sure this is needed. You can set the resource root in `StoreOpt`.

File: src/toml_impl.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
+//! toml universal implementors.
```
Comment (phaazon) [inline src/toml_impl.rs]: TOML

File: src/toml_impl.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+use std::fmt;
+use std::io;
+use std::path::PathBuf;
+
+use crate::key::Key;
+use crate::load::{Load, Loaded, Storage};
+use std::fs::read_to_string;
+use toml::{self, from_str};
+
+/// The Toml universal method. Use this with [`Storage::get_by`] or [`Storage::get_proxied_by`] to
```
Comment (phaazon) [inline src/toml_impl.rs]: TOML

File: src/toml_impl.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+/// benefit from the automatic implementors.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct Toml;
+
+/// Possible error that might occur while loading and reloading TOML formatted scarce resources.
+#[derive(Debug)]
+pub enum TomlError {
+  /// An error in [toml](https://crates.io/crates/toml).
```
Comment (phaazon) [inline src/toml_impl.rs]: TOML. Nice to have put a link to the crate here. 🙂

File: src/toml_impl.rs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        write!(f, "cannot read file {}: {}", path.display(), e)
+      }
+
+      TomlError::NoKey => f.write_str("no path key available"),
+    }
+  }
+}
+
+impl<C, K, T> Load<C, K, Toml> for T
+where
```
Comment (phaazon) [inline src/toml_impl.rs]: Redo the formatting like the json module.

**Comment by phaazon** [comment]
That should be could to go. :)

**Comment by phaazon** [comment]
So basically, if you’re interested, I made the example require the `toml-impl` feature. Everything else is fine, I just renamed / moved things arounde, Thank you so much for your contribution, it’s well appreciated! 😍

**Comment by kampffrosch94** [comment]
Thank you for merging. And for your awesome crate :)

**Pull request merged**